### ISSUE TITLE
Extra bytes when decoding ByteField

### ISSUE THREAD

**Issue by normster** [libffcv/ffcv#73]
Hi,

I'm encoding and storing string fields in my dataset under the ByteField with:

```
title = "denise's peanut chicken"
np.frombuffer(title.encode('utf-8'), dtype=np.uint8)
```

When I decode the same field in the loader, I'm getting a bunch of extra trailing bytes: 

```
>>> titles[0].tobytes().decode('utf-8')
'denise\'s peanut chicken\x00\x08\x00\x00\x00\x00\x00\x00\x00\x1c\x00\x00\x00\x00\x00\x00\x00\x12\x00\x00\x00\x00\x00\x00\x00\x15\x00\x00\x00\x00\x00\x00\x00\n\x00\x00\x00\x00\x00\x00\x00"\x00\x00\x00\x00\x00\x00\x00\n\x00\x00\x00\x00\x00\x00\x00$\x00\x00\x00\x00\x00\x00\x00\x10\x00\x00\x00\x00\x00\x00\x00/\x00\x00\x00\x00\x00\x00\x00\x11\x00\x00\x00\x00\x00\x00\x00\x0e\x00\x00\x00\x00\x00\x00\x00\x0f\x00\x00\x00\x00\x00\x00\x00\x1e\x00\x00\x00\x00\x00\x00\x00\x0f\x00\x00\x00\x00\x00\x00\x00\x16\x00\x00\x00\x00\x00\x00\x00\t\x00\x00\x00\x00\x00\x00\x00-\x00\x00\x00\x00\x00\x00\x00H\x00\x00\x00\x00\x00\x00\x00\x0e\x00\x00\x00\x00\x00\x00\x00$\x00\x00\x00\x00\x00\x00\x00\x1f\x00\x00\x00\x00\x00\x00\x00;\x00\x00\x00\x00\x00\x00'
```

Where do these extra bytes come from?

**Comment by GuillaumeLeclerc** [comment]
Tensors (numpy/pytorch) are dense by default and we are processing batches. There is no way to have one sample with 10 characters and another sample with 15 and have them on the same batch. There is no other solution than doing padding. To my knowledge most neural network take padded text as an input similar to what you are showing. Which model are you using ?

**Comment by andrewilyas** [comment]
hi @normster ! If the strings are relatively small and roughly consistent in size, you might be able to get a bit more performance out of an NDArrayField rather than BytesField. See this comment for a working example of saving and loading variable-length strings with NDArrayField: https://github.com/libffcv/ffcv/issues/57#issuecomment-1016529098

**Comment by normster** [comment]
Thanks for the replies! I understand the need for padding and the tokenizer will be able to pad a shorter string appropriately but I'm confused where the extra non-null bytes are coming from and how I can remove them. Is it because ffcv doesn't zero-out the allocated memory?

@andrewilyas thanks for the suggestion! Does that field pad all entries up to the MAX_STIRNG_SIZE? My current bottleneck is mostly on decoding images but I'll give this a try if I'm still seeing slow data loading.

**Comment by andrewilyas** [comment]
Exactly yeah! It will pad up to MAX_STRING_SIZE with null bytes, so if MAX_STRING_SIZE is something reasonably small I think it might be faster and easier than using Bytes.

**Comment by normster** [comment]
Ah I see. In my case the strings are somewhat inconsistent in length so it might incur a large storage cost unless I set the MAX_STRING_SIZE to something small but I'll keep this in mind :)

**Comment by andrewilyas** [comment]
Makes sense! AFAIK I think that there are no guarantees for the padding of ByteField as of now, so the best thing to do for now would be to store the length of the string as a field and slice it that way.

**Comment by normster** [comment]
Sounds good!

**Comment by andrewilyas** [comment]
Closing this now, feel free to re-open (or open another issue) if anything else comes up!
